### ISSUE TITLE
[Bug] DSV4 indexer: the two torch reference implementations disagree on length masking (latent; consumers currently re-mask)

### ISSUE THREAD

**Issue by Hakureirm** [sgl-project/sglang#33247]
## Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Describe the bug

`indexer.py` carries two pure-torch implementations of the same function, and they **disagree on length masking**. The one selected on every non-SM120 device is the one that does not mask.

```python
elif envs.SGLANG_FP8_PAGED_MQA_LOGITS_TORCH.get():
    if is_sm120_supported():
        fn = fp8_paged_mqa_logits_torch_sm120     # masks correctly
    else:
        fn = fp8_paged_mqa_logits_torch           # does not mask
```

Fed identical inputs (`B=4`, `seq_lens=[64,128,200,256]`, `max_seq_len=256`, `clean_logits=False`):

| | positions beyond `seq_len` | finite entries per row |
|---|---|---|
| `fp8_paged_mqa_logits_torch` | **`0.0` — unmasked** | always 256, regardless of `seq_len` |
| `fp8_paged_mqa_logits_torch_sm120` | `-inf` | exactly `seq_len` |

On the finite entries the two are **bitwise identical** (max abs diff 0). The only difference is the mask: 376 of 1024 positions differ, all of them "should have been masked but was left at 0.0". The row whose `seq_len` equals `max_seq_len` agrees exactly, which confirms the difference is the length mask and nothing else.

## Why it matters

These logits feed a top-k selection of which KV tokens attention will look at. Padding positions sitting at `0.0` compete against real tokens whose logits are negative, so **real tokens get pushed out of the top-k by positions that do not exist**. Nothing crashes; the model just attends to the wrong places.

Both docstrings emphasise CUDA-graph compatibility —

- `fp8_paged_mqa_logits_torch`: *"Vectorized implementation compatible with CUDA graph capture."*
- `fp8_paged_mqa_logits_torch_sm120`: *"CUDA-graph-compatible FP8 paged MQA logits for SM120 (vectorized, no `.item()`)."*

— and **neither mentions that the masking behaviour differs**. Reading the docstrings, they look like two spellings of one function.

## Reproduction

```python
from sglang.srt.layers.attention.dsv4.indexer import (
    fp8_paged_mqa_logits_torch, fp8_paged_mqa_logits_torch_sm120)

# identical q / kvcache / weight / page_table, seq_lens = [64,128,200,256], max_seq_len=256
a = fp8_paged_mqa_logits_torch     (q, kv, w, seq_lens, pt, None, 256, clean_logits=False)
b = fp8_paged_mqa_logits_torch_sm120(q, kv, w, seq_lens, pt, None, 256, clean_logits=False)

torch.isfinite(a).sum(-1)   # tensor([256, 256, 256, 256])   <-- ignores seq_lens
torch.isfinite(b).sum(-1)   # tensor([ 64, 128, 200, 256])   <-- correct
(a[torch.isfinite(a) & torch.isfinite(b)] - b[...]).abs().max()   # 0.0
```

No GPU-hours needed — this runs in seconds.

## Suggested fix

Either apply the same length mask in `fp8_paged_mqa_logits_torch`, or drop the architecture branch entirely and always use the implementation that masks. The second is simpler and removes a class of bug rather than one instance: the branch is selecting between two implementations that are *not* semantically equivalent, so the architecture predicate is silently deciding correctness, not just performance.

If the branch is kept, the docstrings should state the difference explicitly.

## Environment

```
sglang        v0.5.16 (lmsysorg/sglang:latest)
torch/triton  2.11.0+cu130 / 3.6.0
GPU           8x A800-SXM4-80GB (CC 8.0)  — but the bug is device-independent;
              the SM120 branch is simply the only one that avoids it
```

Related: #33194, #33245, #33246.

**Comment by efschu** [comment]
Can confirm the divergence — my fork carries the same file lineage and shows exactly this: the non-SM120 reference masks out-of-length positions to `0.0`, the SM120 one to `-inf`, bitwise identical everywhere else.

One nuance worth pinning before choosing the fix: further down the same file the top-k stage applies its own mask (`masked_scores.masked_fill_(~valid_mask, float("-inf"))`). Whether padding can actually reach the *selected indices* therefore depends on the call path (`clean_logits=False` callers in particular) — I'm running a falsifier on the end-to-end path now and can report back.

Either way, +1 on collapsing the branch rather than patching the leaf: an architecture predicate should select performance, never semantics. The docstrings reading as "two spellings of one function" is how this class of bug survives review.

**Comment by Hakureirm** [comment]
@efschu you're right, and I checked before waiting for your falsifier — **my impact claim in the original post is wrong. Retracting it.**

All three top-k consumers take `seq_lens` and apply their own length mask, so the unmasked `0.0` cannot reach the selected indices:

- `topk_transform_512_pytorch_vectorized` — `indexer.py:280-283`
  ```python
  valid_mask = positions < seq_lens.unsqueeze(1)
  masked_scores.masked_fill_(~valid_mask, float("-inf"))
  ```
- `topk_transform_512` and `topk_transform_512_v2` — `jit_kernel/dsv4/topk.py:45, 86` — both take `seq_lens` as a kernel argument, and the v2 docstring is explicit that length drives the routing (*"A length of 0 is the valid way to express 'no tokens': the row takes the trivial path and the output is all -1"*).

So "real tokens get pushed out of the top-k by positions that do not exist" **does not follow**. I asserted a consequence I had not traced to the consumer, and that was sloppy — the correct statement is only what I actually measured: two functions documented as interchangeable produce bitwise different outputs, and an architecture predicate chooses between them.

**What I still think is worth fixing**, now with the impact stated accurately:

1. The divergence is currently harmless *because every consumer happens to re-mask*. That is a property of today's call sites, not of the function contract — the next consumer that trusts the returned logits inherits a silent bug. The `-inf` version is the one that is safe to trust.
2. The predicate still selects semantics rather than performance, which is the part you put well: *"an architecture predicate should select performance, never semantics."*
3. The docstrings still read as two spellings of one function while differing in masking.

So: same fix (collapse the branch onto the masking implementation), lower severity — a latent contract bug rather than an active correctness bug. I'd rather have this on the record accurately than have the issue oversell itself.

Please do post your end-to-end falsifier result if you've run it — an empirical "no observable difference in selected indices" is a stronger closing statement than my source reading.

**Comment by efschu** [comment]
Falsifier results, as promised. I drove the real call path end-to-end (leaf logits fn → top-k selection) with deliberately unmasked tails — both `0.0` and `1e30` standing in for uninitialized memory: **zero out-of-length positions reach the selected indices.** Two independent reasons: my tree had already collapsed the arch branch onto the masking variant (your "stronger suggestion", implemented a day earlier for an unrelated reason — the unmasked variant also asserts on the 2-D `seq_lens` the paged call site passes), and all three top-k consumers bound by `seq_len` independently (both CUDA kernels never load past the end; the torch fallback re-masks with `-inf`).

So on a tree with the branch collapsed this is not a serving-correctness bug — but two things are worth fixing beyond the collapse. First, the practical defect is the **test oracle**: the unmasked function is the reference the paged-MQA parity test validates against, and the compare helper skips exactly the region where the two disagree — so the oracle can't catch a masking regression. Second, the honest contract is that **the logits tail is undefined, not `-inf`**: tilelang leaves the trailing partial 64-block as GEMM output over stale cache slots, aiter hands the kernel a `torch.empty`. Consumers must bound by `seq_len`; `-inf` is not a guarantee to code against, and documenting that prevents the next variant of this bug.

Fix on my side: reference aligned to `-inf`, compare helper now asserts `-inf` on both sides, docstrings state masking and reachability, and the seq_len-bounding contract is pinned at the top-k dispatch with regression tests.

**Comment by Hakureirm** [comment]
Thanks for running the falsifier — that closes the serving-correctness question properly, and it is stronger evidence than my source reading was.

I checked both of your follow-up claims against the tree rather than taking them on trust. **One holds, one does not.**

### Holds: the contract is "tail undefined", not `-inf`

Upstream already says so, in the test itself — `test/registered/kernels/test_sm120_paged_mqa_logits.py:141`:

```python
"""Compare reference (uninitialized beyond seq_len) vs SM120 (-inf beyond)."""
```

That is worth pinning, and it makes my original framing weaker than my retraction admitted: I described these as "two functions documented as interchangeable", but the test **documents the divergence explicitly**. The docstrings on the two functions don't — that gap is real, and it is the part actually worth fixing.

### Does not hold: the oracle is not blind to a masking regression

You wrote that the compare helper *"skips exactly the region where the two disagree — so the oracle can't catch a masking regression."* The helper does two things, not one (`:143-155`):

```python
    # Valid positions must match
    for i in range(batch_size):
        sl = int(seq_lens[i].item())
        torch.testing.assert_close(ref[i, :sl], sm120[i, :sl], atol=atol, rtol=rtol, ...)
    # Invalid positions in SM120 output must be -inf
    positions = torch.arange(max_seq_len, device=sm120.device)
    invalid = positions.unsqueeze(0) >= seq_lens.unsqueeze(1)
    assert torch.all(
        torch.isinf(sm120[invalid]) & (sm120[invalid] < 0)
    ), "SM120 output must fill invalid positions with -inf"
```

The second assertion covers exactly the masking behaviour. If the SM120 variant stopped writing `-inf` past `seq_len`, this test fails. What the oracle deliberately does not check is the **reference's** tail — correctly, because that region is undefined by contract, which is your own point.

So the oracle is doing the right thing on both sides. If it should be strengthened, the target is not the skip, it is that the assertion only binds the SM120 side while the seq_len-bounding requirement it implies is nowhere stated in the functions' own docstrings.

### Where that leaves this issue

Netting out what survived verification:

- **Not a serving-correctness bug** — your falsifier, plus all three top-k consumers bounding by `seq_len`.
- **Not an undocumented divergence** — the test documents it; only the two docstrings omit it.
- **Not an untested behaviour** — the masking side is asserted.

That is thin enough that I'd rather this be closed as working-as-intended than left open implying a defect. The residue worth keeping is documentation: state on both functions that **the logits tail beyond `seq_len` is undefined and consumers must bound by `seq_len`**, since as you note tilelang leaves GEMM output over stale slots and aiter passes `torch.empty` — so `-inf` is an implementation detail of one variant, not a contract.

I'll leave it open briefly in case a maintainer wants the docstring change tracked here; happy to close otherwise. Apologies for the noise — I filed this asserting a consequence I had not traced to the consumer, which is on me.

**Comment by efschu** [comment]
You're right, and the correction stands for my tree as well: I re-checked the pre-fix state, and `_compare` carried the same second assertion (SM120 tail must be `-inf`) plus the docstring documenting the divergence all along. My "the oracle can't catch a masking regression" claim was wrong — the helper binds the SM120 side correctly and deliberately leaves the reference tail unchecked, which is consistent with the tail-undefined contract. I should have verified that against the file instead of restating it.

Agreed on the net-out: what's left is exactly the function docstrings omitting what the test already documents (fixed on my side), plus — as hygiene rather than correctness — aligning the unreachable reference to `-inf` and collapsing the arch branch. Thanks for checking both claims against the tree; that's the review this deserved.

**Comment by efschu** [comment]
Opened https://github.com/sgl-project/sglang/pull/33260. It is narrower than the original report and follows where this thread landed: the substantive part is that the `is_sm120_supported()` split hands non-SM120 devices `fp8_paged_mqa_logits_torch`, which asserts on a 1-D `seq_lens` while the paged call site unsqueezes to 2-D for exactly this backend — so `SGLANG_FP8_PAGED_MQA_LOGITS_TORCH=1` raises `AssertionError` there rather than mis-masking. Collapsing onto the masking variant fixes that. The `-inf` alignment and the docstrings stating the tail-undefined contract ride along as the documentation residue you identified; I did not touch the parity test, since — as established above — its second assertion already binds the SM120 side correctly.

**Comment by Hakureirm** [comment]
Closing: this was reported while working around a wrong assumption on my side.

DeepSeek-V4 does load and serve on Ampere with the stock checkpoint — SGLang already routes its fp4 experts to a Marlin kernel that has no architectural need for SM90. Once that single gate is relaxed the model runs (verified on 8x A800: needle 6/6, 51.9 tok/s). The blockers described here were all downstream of repacking the checkpoint into a layout it did not need to be in.

Upstream has said it does not plan to support SM80, which is a reasonable scope decision, so I am not filing a patch for the gate. Leaving this note in case someone hits the same wall.
